core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -449,9 +448,16 @@ private[log] class Cleaner(val id: Int,
                              map: OffsetMap,
                              retainDeletes: Boolean,
                              maxLogMessageSize: Int,
+                             activePids: Map[Long, ProducerIdEntry],
                              stats: CleanerStats) {
     val logCleanerFilter = new RecordFilter {
-      def shouldRetain(record: Record): Boolean = shouldRetainMessage(source, map, retainDeletes, record, stats)
+      def shouldRetain(recordBatch: RecordBatch, record: Record): Boolean = {
+        // retain the entry if it is the last one produced by an active idempotent producer to ensure that
+        // the PID is not removed from the log before it has been expired
+        val pid = recordBatch.producerId
+        val isLastEntryForPid = RecordBatch.NO_PRODUCER_ID < pid && activePids.get(pid).exists(_.lastOffset == record.offset)
+        record.isControlRecord || isLastEntryForPid || shouldRetainMessage(source, map, retainDeletes, record, stats)
```

Comment:
I don't think we should split the logic between this and `shouldRetainMessage`. For example, the latter already checks `record.isControlRecord`.